arch/x86/entry.S

```diff
@@ -32,38 +32,46 @@
 #include <usermode.h>
 #include <errno.h>
 
-.macro _handle_usermode cr3
+.macro _from_usermode switch_stack=1
+    SET_CR3 cr3
+    swapgs
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+.endm
+
+.macro _to_usermode switch_stack=1
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+    swapgs
+    SET_CR3 user_cr3
+.endm
+
+.macro exc_from_usermode
     testb $0x3, cpu_exc_cs(%_ASM_SP)
     jz 1f /* skip if from kernel mode */
-        SET_CR3 \cr3
-        swapgs
+        _from_usermode switch_stack=0
     1:
 .endm
 
-.macro enter_from_usermode
-    _handle_usermode cr3
-.endm
-
-.macro enter_to_usermode
-    _handle_usermode user_cr3
+.macro exc_to_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _to_usermode switch_stack=0
+    1:
```

Comment:
Isn't this always executed in kernel mode? Again does it make sense to hoist the conditional?

Reply:
Yes. We have to be sure whether we did the transition earlier or not. Will apply some abstraction to it.